sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/GpuIntervalUtils.scala

```diff
@@ -0,0 +1,556 @@
+      withResource(Scalar.fromNull(DType.INT64)) { nullScalar =>
+        // if overflow, set as null
+        overflow.ifElse(nullScalar, ret)
+      }
+    }
+  }
+
+  // Add left and right to a new one and then close them.
+  // Check overflow and set null
+  private def add(left: ColumnVector, right: ColumnVector): ColumnVector = {
+    withResource(left) { l =>
+      withResource(right) { r =>
+        withResource(l.add(r)) { result =>
+          setNullIfOverflow(l, r, result)
+        }
+      }
+    }
+  }
+
+  private def daysToMicros(
+      sign: ColumnVector, daysInGroupTable: ColumnVector, maxDay: Long): ColumnVector = {
+    multiple(sign, daysInGroupTable, DAYS.toMicros(1), maxDay)
+  }
+
+  private def hoursToMicros(
+      sign: ColumnVector, hoursInGroupTable: ColumnVector, maxHour: Long): ColumnVector = {
+    multiple(sign, hoursInGroupTable, HOURS.toMicros(1), maxHour)
+  }
+
+  private def minutesToMicros(
+      sign: ColumnVector, minutesInGroupTable: ColumnVector, maxMinute: Long): ColumnVector = {
+    multiple(sign, minutesInGroupTable, MINUTES.toMicros(1), maxMinute)
+  }
+
+  private def secondsToMicros(
+      sign: ColumnVector, secondsInGroupTable: ColumnVector, maxSecond: Long): ColumnVector = {
+    multiple(sign, secondsInGroupTable, SECONDS.toMicros(1), maxSecond)
+  }
+
+  private def secondsToMicros(
+      sign: ColumnVector, secondsInGroupTable: ColumnVector): ColumnVector = {
+    multiple(sign, secondsInGroupTable, SECONDS.toMicros(1))
+  }
+
+  /**
+   * Check range, return sign * base * multiple.
+   *
+   * @param sign      long column with 1L or -1L in it
+   * @param base      string column contains positive long
+   * @param multiple  const long value
+   * @param maxInBase the max value for base column
+   * @return
+   */
+  private def multiple(
+      sign: ColumnVector, base: ColumnVector, multiple: Long, maxInBase: Long): ColumnVector = {
+    withResource(base.castTo(DType.INT64)) { baseLong =>
+      withResource(Scalar.fromLong(maxInBase)) { maxScalar =>
+        withResource(baseLong.greaterThan(maxScalar)) { greater =>
+          withResource(Scalar.fromNull(DType.INT64)) { nullScalar =>
+            withResource(greater.ifElse(nullScalar, baseLong)) { baseWithFix =>
+              withResource(baseWithFix.mul(sign)) { baseWithSign =>
+                withResource(Scalar.fromLong(multiple)) { multipleScalar =>
+                  baseWithSign.mul(multipleScalar)
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Return sign * base * multiple.
+   *
+   * @param sign         long column
+   * @param groupInTable string column contains positive long
+   * @param multiple     const long value
+   * @return sign * base * multiple
+   */
+  private def multiple(
+      sign: ColumnVector, groupInTable: ColumnVector, multiple: Long): ColumnVector = {
+    withResource(groupInTable.castTo(DType.INT64)) { baseLong =>
```

Comment:
`baseLong` can be closed right away after getting `baseWithSign`.
One more thing, could we only compute the `baseLong.mul(sign)` when the `sign` is `-1`. For sign being `+1`, we could just increase the reference count of `baseLong` by 1.

Reply:
>baseLong can be closed right away after getting baseWithSign

Yes.

> For sign being +1, we could just increase the reference count of baseLong by 1. Doing this may get a better perf for positive sign.

The `sign` is a column with -1 or 1 in it, it's not a scalar.